astropy/tests/pytest_plugins.py

```diff
@@ -479,10 +487,23 @@ def pytest_runtest_setup(item):
     if item.config.getvalue('open_files'):
         item.open_files = _get_open_file_list()
 
-    if ('remote_data' in item.keywords and
-            not item.config.getvalue("remote_data")):
-        pytest.skip("need --remote-data option to run")
+    remote_data = item.keywords.get('remote_data')
+
+    remote_data_config = item.config.getvalue("remote_data")
+
+    if remote_data is not None:
+
+        source = remote_data.kwargs.get('source', 'any')
+
+        if source not in ('astropy', 'any'):
+            raise ValueError("source should be 'astropy' or 'any'")
 
+        if remote_data_config == 'none':
+            if source in ('astropy', 'any'):
```

Comment:
This is always true because you already raised an exception `if source not in ('astropy', 'any')`.